react_basic/src/pages/PageWrapper.jsx

```diff
@@ -0,0 +1,154 @@
+      if (e.key === 'B' && e.shiftKey && e.ctrlKey && e.altKey) {
+        e.stopImmediatePropagation();
+
+        navigate('/billy');
+      }
+    });
+  }, []);
+
+  function successPopupResetCallback() {
+    setSuccessMsg();
+    setSuccessMsgCallback();
+    setSuccessMsgMillis();
+  }
+
+  function errorPopupResetCallback() {
+    setErrorMsg();
+    setErrorMsgCallback();
+    setErrorMsgMillis();
+  }
+
+  useEffect(() => {
+    if (location.pathname === '/') {
+      navigate('/blog');
+    } else if (location.pathname === '/blog') {
+      const query = localStorageService.getItemForUser(user, 'query');
+      const sort = localStorageService.getItemForUser(user, 'sort');
+      const order = localStorageService.getItemForUser(user, 'order');
+
+      urlUtils.changeQuery({
+        query: {
+          operation: query ? 'set' : 'delete',
+          value: query,
+        },
+        sort: {
+          operation: sort ? 'set' : 'delete',
+          value: sort,
+        },
+        order: {
+          operation: order ? 'set' : 'delete',
+          value: order,
+        },
+      });
+    }
+  }, [location]);
+
+  useEffect(() => {
+    localStorageService.setLoggedInUser(user);
+    setTheme(themeService.getThemeForUser(user));
+    setFavorites(user?.favorites ? user.favorites : []);
+  }, [user]);
+
+  useEffect(() => {
+    themeService.applyTheme(user, theme);
+  }, [theme]);
+
+  return (
+    <PageWrapperContext.Provider
+      value={{
+        user,
+        setUser,
+        theme,
+        setTheme,
+        favorites,
+        setFavorites,
+        setSuccessMsg,
+        setSuccessMsgCallback,
+        setSuccessMsgMillis,
+        setErrorMsg,
+        setErrorMsgCallback,
+        setErrorMsgMillis,
+        setFullscreenPopupContent,
+      }}
+    >
+      {successMsg && (
+        <PopupMessage
+          key={'success-msg-' + new Date().getTime()}
+          level="success"
+          message={successMsg}
+          resetCallback={successPopupResetCallback}
+          milliseconds={successMsgMillis}
+          callback={successMsgCallback}
+        />
+      )}
+
+      {errorMsg && (
+        <PopupMessage
+          key={'error-msg-' + new Date().getTime()}
+          level="error"
+          message={errorMsg}
+          resetCallback={errorPopupResetCallback}
+          milliseconds={errorMsgMillis}
+          callback={errorMsgCallback}
+        />
+      )}
+
+      {fullscreenPopupContent && (
+        <FullscreenPopup setFullscreenPopupContent={setFullscreenPopupContent}>
+          {fullscreenPopupContent}
+        </FullscreenPopup>
+      )}
+
+      <section className="page-wrapper">
+        <Header />
+        <div className="page-content-wrapper">
+          <Advert />
+          <main className="page-main">
+            <ErrorBoundary fallback={<ErrorMsg />}>
+              <Outlet />
+            </ErrorBoundary>
+          </main>
+          <Footer />
+        </div>
+      </section>
```

Comment:
And for this usually created separate component Layout, and in the end of refactoring we will have structured easy to understand component:
![image](https://github.com/user-attachments/assets/5aa3c785-4dad-4249-b652-cebfe0ebdacc)